### PR TITLE
Show warning message when pressure is high and no correction is selected 118360565

### PR DESCRIPTION
**Pull request by brennanneoh**
@rpbaltazar Have a look

### CODE REVIEW

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 1]
```diff
@@ -44,18 +44,42 @@ describe 'orifice-calculator-viewmodel-spec', ->
...
+    describe 'when pressure is high', ->
+      it 'should be invalid at > 200', ->
+        viewModel.operatingPressure(300)
+        expect(viewModel.operatingPressure.isValid()).toEqual false
+        viewModel.operatingPressure(1000)
+        expect(viewModel.operatingPressure.isValid()).toEqual false
+
+      it 'should display a warning message at > 200', ->
+        viewModel.operatingPressure(201)
+        expect(viewModel.operatingPressure.error()).toEqual OPL.OrificeCalculator.Config.Dictionaries.Messages.warning
```
Comment (rpbaltazar) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: 1- Don't rely on observable validations for error message. The way I see it an error is something that stops the user from completing a task, which is not the case. Rather use a pureComputed to decide which feedback to give to the user
2- `OPL.OrificeCalculator.Config.Dictionaries.Messages.warning` is very vague. If I show you this variable name without any usage background you have no clue of what will be translated to. Please add at least a reference to the operating pressure. Something along `OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarning`

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 2]
```diff
@@ -44,18 +44,42 @@ describe 'orifice-calculator-viewmodel-spec', ->
       expect(viewModel.pipeID().length).toEqual 6
 
     it 'should have default value of "1.939\'\' XS, Sch 80, Sch 80S"', ->
-      expect(viewModel.selectedPipeID()).toEqual '1.939\'\' XS, Sch 80, Sch 80S'
+      expect(viewModel.selectedPipeID()).toEqual OPL.OrificeCalculator.Config.Dictionaries.PipeID.one_nine_inch
```
Comment (rpbaltazar) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: I reckon it might have been me the first one to write it this way, but JS variables use camel case:
`OPL.OrificeCalculator.Config.Dictionaries.PipeID.oneNineInch`

Please fix the other keys

Comment (brennanneoh) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Done

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee (hunk 1 above)
Comment (brennanneoh) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Done

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 3]
```diff
@@ -44,18 +44,38 @@ describe 'orifice-calculator-viewmodel-spec', ->
...
+        expect(viewModel.operatingPressureWarningMessage()).toEqual OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
+
+      it 'should display an operatingPressureError message at > 400', ->
+        viewModel.operatingPressure(401)
+        expect(viewModel.operatingPressureWarningMessage()).toEqual OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
+
+    describe 'when pressure is low', ->
+      it 'should not have any messages', ->
+        viewModel.operatingPressure(100)
+        expect(viewModel.operatingPressure.isValid()).toEqual true
```
Comment (rpbaltazar) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: isValid is no longer being used. Check that the operatingPressureWarningMessage is not defined

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 4]
```diff
@@ -44,18 +44,38 @@ describe 'orifice-calculator-viewmodel-spec', ->
       expect(viewModel.pipeID().length).toEqual 6
 
     it 'should have default value of "1.939\'\' XS, Sch 80, Sch 80S"', ->
-      expect(viewModel.selectedPipeID()).toEqual '1.939\'\' XS, Sch 80, Sch 80S'
+      expect(viewModel.selectedPipeID()).toEqual OPL.OrificeCalculator.Config.Dictionaries.PipeID.oneNineInch
```
Comment (Ranhiru) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: `OneNineInch` ?

File: src/js/ko-components/templates/orifice-calculator.haml [hunk 5]
```diff
@@ -39,3 +39,4 @@
         %div
           %input.compressibility-correction{ :type => "radio", "data-bind" => "checkedValue: $data, checked: $component.chosenCompressibilityCorrection"  }
           %label{ "data-bind" => "text: $data" }
+      %span{ "data-bind" => "visible: operatingPressure() > 200, text: operatingPressureWarningMessage" }
```
Comment (rpbaltazar) [inline src/js/ko-components/templates/orifice-calculator.haml]: Instead of having magic numbers on the template, rely on the operatingPressureWarningMessage being defined or not.

`%span{ "data-bind" => "visible: operatingPressureWarningMessage, text: operatingPressureWarningMessage" }`

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 6]
```diff
@@ -5,40 +5,64 @@
...
+    @selectedPipeID = ko.observable OPL.OrificeCalculator.Config.Dictionaries.PipeID.oneNineInch
+    @operatingPressure = ko.observable(0).extend
+      required:
+        params: true
+        message: OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureError
+      digit:
+        params: true
+        message: OPL.OrificeCalculator.Config.Dictionaries.Messages.integerError
+
+    @operatingPressureWarningMessage = ko.pureComputed (->
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: You can use the fat arrow ( => ) to bind this (@) inside the function. Reads better and is the pattern we use.

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee [hunk 7]
```diff
@@ -44,18 +44,38 @@ describe 'orifice-calculator-viewmodel-spec', ->
...
 
   describe 'operatingPressure', ->
     itBehavesLikeMandatoryField viewModel.operatingPressure
     itBehavesLikeIntegerField viewModel.operatingPressure
 
+    describe 'when pressure is high', ->
+
+      it 'should display a operatingPressureWarning message at > 200', ->
+        viewModel.operatingPressure(201)
+        viewModel.operatingPressure.valueHasMutated()
```
Comment (Ranhiru) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Why is this required here ? I don't see it used in other specs

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 8]
```diff
@@ -5,40 +5,64 @@
...
+      if (@operatingPressure() > 200 and @operatingPressure() < 401)
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Please avoid using if else chained. there is another form that looks so much nicer: http://coffeescript.org/#switch

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 9]
```diff
@@ -5,40 +5,64 @@
...
+        return OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
+      else if (@operatingPressure() > 400)
+        return OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
+      else
+        return ''
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: returning undefined instead of '' will make you have one less logical operation on the template. Check above comment (on template file)

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 10]
```diff
@@ -5,40 +5,64 @@
...
+      ), @
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Therefore this line is not needed

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee (hunk 3 above)
Comment (brennanneoh) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Done

File: src/js/ko-components/templates/orifice-calculator.haml (hunk 5 above)
Comment (brennanneoh) [inline src/js/ko-components/templates/orifice-calculator.haml]: Done

File: spec/js/ko-components/view-models/orifice-calculator-spec.coffee (hunk 7 above)
Comment (brennanneoh) [inline spec/js/ko-components/view-models/orifice-calculator-spec.coffee]: Forgot to remove this.

File: src/js/ko-components/view-models/orifice-calculator.coffee (hunk 6 above)
Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Done.

File: src/js/ko-components/view-models/orifice-calculator.coffee (hunk 8 above)
Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Done

File: src/js/ko-components/view-models/orifice-calculator.coffee (hunk 9 above)
Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Done

File: src/js/ko-components/view-models/orifice-calculator.coffee (hunk 10 above)
Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Done

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 11]
```diff
@@ -14,14 +14,12 @@ class OPL.KoComponents.ViewModels.OrificeCalculator
...
-    @operatingPressureWarningMessage = ko.pureComputed (->
-      if (@operatingPressure() > 200 and @operatingPressure() < 401)
-        return OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
-      else if (@operatingPressure() > 400)
-        return OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
-      else
-        return ''
-      ), @
+    @operatingPressureWarningMessage = ko.pureComputed =>
+      switch true
```
Comment (Ranhiru) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: I'm not very fond of the `switch true` although I understand why it's used here. You could alternatively assign the value to a var and return that.

```
grade = switch
  when score < 60 then 'F'
  when score < 70 then 'D'
  when score < 80 then 'C'
  when score < 90 then 'B'
  else 'A'
return grade
```

What do you think @rpbaltazar  ?

File: src/js/ko-components/view-models/orifice-calculator.coffee [hunk 12]
```diff
@@ -5,40 +5,62 @@
...
+    @operatingPressure = ko.observable(0).extend
+      required:
+        params: true
+        message: OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureError
+      digit:
+        params: true
+        message: OPL.OrificeCalculator.Config.Dictionaries.Messages.integerError
+
+    @operatingPressureWarningMessage = ko.pureComputed =>
+      switch true
```
Comment (rpbaltazar) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Grabbing @Ranhiru's comment and adjusting the code to the situation here:

```
switch
  when @operatingPressure() < 201 then undefined
  when 201 <= @operatingPressure() < 400 then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
  when 401 <= @operatingPressure() then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
```

Comment (brennanneoh) [inline src/js/ko-components/view-models/orifice-calculator.coffee]: Done. We did this though:
```
switch
  when 201 <= @operatingPressure() < 400 then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningMayResult
  when 401 <= @operatingPressure() then OPL.OrificeCalculator.Config.Dictionaries.Messages.operatingPressureWarningWillResult
```

**Comment by brennanneoh** [comment]
Merged by @samkahchiin  and @brennanneoh.
